File: idp-portal/backend/app/models/catalog.py
"""Catalog models for Software Catalog actions (Story 2.1, FR1; Story 2.6 tags).

Defines Pydantic models for:
- ActionEngine, ActionPlatform, ActionStatus enums
- ActionCreate: input model for creating actions
- ActionResponse: output model for action list
- ActionDetail: output model with full details (Story 2.14: rbac_policies removed — RBAC via profiles)
- TagCreate, TagResponse: tag models (Story 2.6, FR11c)
- Story 2.23: ActionCategory removed — use tags for categorization
"""

from datetime import datetime
from enum import Enum
from typing import Any

import re

from pydantic import BaseModel, Field, field_validator, model_validator


class ActionEngine(str, Enum):
    """Supported database engines."""
    ORACLE = "Oracle"
    SQL_SERVER = "SQL Server"
    DB2 = "DB2"


class ActionPlatform(str, Enum):
    """Execution platforms for actions."""
    AAP = "AAP"
    GITHUB_ACTIONS = "GitHub Actions"
    AZURE_DEVOPS = "Azure DevOps"
    TERRAFORM = "Terraform"


class ActionStatus(str, Enum):
    """Action lifecycle status."""
    DRAFT = "draft"
    PUBLISHED = "published"
    DISABLED = "disabled"


class ImpactLevel(str, Enum):
    """Impact levels for actions per environment."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


# === Story 2.6: Tag models (FR11c) ===


def normalize_tag_name(name: str) -> str:
    """Normalize tag name: lowercase, strip, replace spaces with nothing."""
    if not name or not isinstance(name, str):
        return ""
    return name.strip().lower().replace(" ", "")


class TagCreate(BaseModel):
    """Input model for creating a tag (Story 2.6, AC #5).

    Attributes:
        name: Tag name (normalized: lowercase, no spaces)
    """
    name: str = Field(..., min_length=1, max_length=255)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, v: str) -> str:
        """Normalize tag name: lowercase, strip, no spaces."""
        normalized = normalize_tag_name(v)
        if not normalized:
            raise ValueError("tag name cannot be empty or whitespace only")
        return normalized


class TagResponse(BaseModel):
    """Output model for tag (Story 2.6, AC #5)."""
    id: int
    name: str
    created_at: datetime


class ActionTagsUpdateRequest(BaseModel):
    """Input for PUT /admin/actions/{id}/tags (Story 2.6, AC #5).

    Provide either tag_ids or tag_names. tag_names creates missing tags on the fly.
    """
    tag_ids: list[int] | None = None
    tag_names: list[str] | None = None

    @field_validator("tag_names")
    @classmethod
    def normalize_tag_names(cls, v: list[str] | None) -> list[str] | None:
        if v is None:
            return None
        out = [normalize_tag_name(s) for s in v if normalize_tag_name(s)]
        return out  # keep [] to allow "clear all tags"

    @model_validator(mode="after")
    def require_tag_ids_or_tag_names_exclusive(self) -> "ActionTagsUpdateRequest":
        if self.tag_ids is None and self.tag_names is None:
            raise ValueError("provide either tag_ids or tag_names")
        if self.tag_ids is not None and self.tag_names is not None:
            raise ValueError("provide either tag_ids or tag_names, not both")
        return self


class ActionCreate(BaseModel):
    """Input model for creating a new action (AC #2, #5).

    Attributes:
        name: Action name (1-255 chars, unique)
        description: Action description (max 4000 chars)
        engine: Database engine (Oracle, SQL Server, DB2)
        platform: Execution platform (AAP, GitHub Actions, Azure DevOps, Terraform)
        parameters_schema: Optional JSON Schema for action parameters
        impact_rules: Optional impact rules per environment
        default_impact_level: Default impact when no rule matches environment (Story 2.18 AC5)
        documentation_md: Optional Markdown documentation (Story 3.4, FR12)
        (Story 2.24: change_model_code removed; change_type_config is in ExecutionStepsUpdate only.)
    """
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = Field(None, max_length=4000)
    engine: ActionEngine
    platform: ActionPlatform
    parameters_schema: dict[str, Any] | None = None
    impact_rules: dict[str, Any] | None = None
    default_impact_level: ImpactLevel | None = None
    documentation_md: str | None = Field(None, max_length=100_000)

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str) -> str:
        """Strip whitespace from name and validate not empty."""
        stripped = v.strip()
        if not stripped:
            raise ValueError("name cannot be empty or whitespace only")
        return stripped

    @field_validator("parameters_schema")
    @classmethod
    def validate_parameters_schema(cls, v: dict[str, Any] | None) -> dict[str, Any] | None:
        """Validate parameters_schema is a valid JSON Schema structure."""
        if v is None:
            return None

        # Basic JSON Schema validation - must be an object with valid structure
        if not isinstance(v, dict):
            raise ValueError("parameters_schema must be a JSON object")

        # Check for basic JSON Schema properties if present
        valid_top_level = {"$schema", "type", "properties", "required", "additionalProperties",
                          "title", "description", "definitions", "$defs", "allOf", "anyOf", "oneOf"}
        if not v or not any(k in valid_top_level for k in v.keys()):
            raise ValueError("parameters_schema must be a valid JSON Schema with recognized properties")

        return v

    @field_validator("impact_rules")
    @classmethod
    def validate_impact_rules(cls, v: dict[str, Any] | None) -> dict[str, Any] | None:
        """Validate impact_rules structure: {environment: {level: "low"|"medium"|"high"|"critical"}}."""
        if v is None:
            return None

        if not isinstance(v, dict):
            raise ValueError("impact_rules must be a JSON object")

        valid_levels = {"low", "medium", "high", "critical"}

        for env, rules in v.items():
            if not isinstance(rules, dict):
                raise ValueError(f"impact_rules[{env}] must be an object with 'level' key")

            level = rules.get("level")
            if level is None:
                raise ValueError(f"impact_rules[{env}] must have a 'level' key")

            if level not in valid_levels:
                raise ValueError(
                    f"impact_rules[{env}].level must be one of: {', '.join(valid_levels)}"
                )

        return v

class ActionResponse(BaseModel):
    """Output model for action in list views (AC #5).

    Excludes rbac_policies for list performance.
    Story 2.6: includes tags (tag names) for display.
    Story 2.18 AC5: includes default_impact_level.
    Story 2.24: change_model_code removed; change_type_config per env.
    Story 2.23: category removed — use tags instead.
    Story 3.4: includes documentation_md (FR12).
    """
    id: int
    name: str
    description: str | None = None
    engine: ActionEngine
    platform: ActionPlatform
    parameters_schema: dict[str, Any] | None = None
    impact_rules: dict[str, Any] | None = None
    default_impact_level: ImpactLevel | None = None
    status: ActionStatus
    created_by: int | None = None
    created_at: datetime
    updated_at: datetime | None = None
    tags: list[str] = Field(default_factory=list)
    documentation_md: str | None = None


class ActionDetail(ActionResponse):
    """Output model for action detail view (AC #5).

    Story 2.14: rbac_policies removed — RBAC now managed via profiles.
    Story 2.24: change_type_config is dict[str, ChangeTypeConfigEntry].
    """
    execution_steps: list["ExecutionStep"] | None = None
    change_type_config: dict[str, "ChangeTypeConfigEntry"] | None = None


# === Story 2.2: Execution Steps and Change Type Models ===
# === Story 2.7: ConnectorType generic connector (AC1, AC4) ===


class ConnectorType(str, Enum):
    """Generic connector types for execution steps (Story 2.7, AC1).

    Replaces is_servicenow_change flag. connector_config holds connector-specific params
    (e.g. ServiceNow change template). conditional_environments applies when step is conditional.
    """
    AAP = "aap"
    SERVICENOW = "servicenow"
    AZUREDEVOPS = "azuredevops"
    JIRA = "jira"
    GITHUB_ACTIONS = "github_actions"
    TERRAFORM = "terraform"
    NONE = "none"


class ExecutionStepType(str, Enum):
    """Types of execution steps (FR2)."""
    PREREQUISITE = "prerequisite"
    EXECUTION = "execution"
    VERIFICATION = "verification"


class ChangeTypeConfigEntry(BaseModel):
    """Per-environment change config (Story 2.24). required=True implies change_model_code required, alphanumeric max 50."""
    required: bool = False
    change_model_code: str | None = Field(None, max_length=50)

    @model_validator(mode="after")
    def require_code_when_required(self) -> "ChangeTypeConfigEntry":
        if self.required and (self.change_model_code is None or not self.change_model_code.strip()):
            raise ValueError("change_model_code is required when required is true")
        if self.change_model_code is not None and self.change_model_code.strip():
            if not re.match(r"^[A-Za-z0-9]+$", self.change_model_code):
                raise ValueError("change_model_code must be alphanumeric only")
        return self


class ExecutionStep(BaseModel):
    """Single execution step configuration (AC #1, #2; Story 2.7: connector_type).

    Attributes:
        order: Step order (1-based, sequential)
        name: Step name (1-255 chars)
        type: Step type (prerequisite, execution, verification)
        connector_type: Generic connector (aap, servicenow, azuredevops, jira, github_actions, terraform, none).
        connector_config: Connector-specific config (e.g. ServiceNow change template). Optional.
        conditional_environments: Required when connector_type is servicenow (environments where step applies).
    """
    order: int = Field(..., ge=1)
    name: str = Field(..., min_length=1, max_length=255)
    type: ExecutionStepType
    connector_type: ConnectorType = ConnectorType.NONE
    connector_config: dict[str, Any] | None = None
    conditional_environments: list[str] | None = None

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str) -> str:
        """Strip whitespace from name and validate not empty."""
        stripped = v.strip()
        if not stripped:
            raise ValueError("step name cannot be empty or whitespace only")
        return stripped

    @field_validator("conditional_environments")
    @classmethod
    def validate_conditional_environments(
        cls, v: list[str] | None, info
    ) -> list[str] | None:
        """Validate conditional_environments is required when connector_type is servicenow (Story 2.7)."""
        connector_type = info.data.get("connector_type", ConnectorType.NONE)
        if connector_type == ConnectorType.SERVICENOW and (v is None or len(v) == 0):
            raise ValueError(
                "conditional_environments is required when connector_type is servicenow"
            )
        return v


class ExecutionStepsUpdate(BaseModel):
    """Input model for updating execution steps (AC #5).

    Attributes:
        steps: Ordered list of execution steps
        change_type_config: Per-env change config (Story 2.24: required + change_model_code)
    """
    steps: list[ExecutionStep]
    change_type_config: dict[str, ChangeTypeConfigEntry] | None = None

    @field_validator("change_type_config", mode="before")
    @classmethod
    def reject_legacy_change_type_config(
        cls, v: dict[str, Any] | None
    ) -> dict[str, Any] | None:
        """Reject legacy format (env -> string) with clear message (Story 2.24 AC4)."""
        if v is None:
            return None
        for env, val in v.items():
            if isinstance(val, str):
                raise ValueError(
                    "change_type_config uses legacy format (environment -> string). "
                    'Use new format: {"ENV": {"required": true|false, "change_model_code": "..."}}.'
                )
        return v

    @field_validator("steps")
    @classmethod
    def validate_steps_order(cls, v: list[ExecutionStep]) -> list[ExecutionStep]:
        """Validate steps order is unique and sequential starting from 1."""
        if not v:
            raise ValueError("at least one step is required")

        orders = [step.order for step in v]

        # Check uniqueness
        if len(orders) != len(set(orders)):
            raise ValueError("step order values must be unique")

        # Check sequential starting from 1
        expected = list(range(1, len(v) + 1))
        if sorted(orders) != expected:
            raise ValueError(
                f"step order must be sequential starting from 1 (expected {expected}, got {sorted(orders)})"
            )

        return v


# === Story 2.4: Status Transition and Lifecycle Models ===
# Note: Story 2.3 RBAC by action models (UserProfile, EnvironmentPermission, RbacPolicies,
# RbacPoliciesUpdate) removed in Story 2.14 — RBAC now managed via profiles.


class StatusTransition(str, Enum):
    """Valid status transitions for action lifecycle (Story 2.4, AC #1, #4, #5).

    State machine:
        draft -> published (publish)
        published -> disabled (disable)
        disabled -> published (enable)
    """
    PUBLISH = "publish"
    DISABLE = "disable"
    ENABLE = "enable"


class InvalidTransitionError(Exception):
    """Raised when an invalid status transition is attempted."""

    def __init__(self, current_status: str, transition: str, message: str | None = None):
        self.current_status = current_status
        self.transition = transition
        if message is None:
            message = f"Transition de statut invalide: {current_status} avec {transition}"
        super().__init__(message)


# Valid transitions mapping: {current_status: {allowed_transition: new_status}}
_VALID_TRANSITIONS: dict[ActionStatus, dict[StatusTransition, ActionStatus]] = {
    ActionStatus.DRAFT: {
        StatusTransition.PUBLISH: ActionStatus.PUBLISHED,
    },
    ActionStatus.PUBLISHED: {
        StatusTransition.DISABLE: ActionStatus.DISABLED,
    },
    ActionStatus.DISABLED: {
        StatusTransition.ENABLE: ActionStatus.PUBLISHED,
    },
}


def validate_transition(current_status: ActionStatus, transition: StatusTransition) -> ActionStatus:
    """Validate and return the new status for a given transition.

    Args:
        current_status: Current action status
        transition: Requested transition

    Returns:
        The new ActionStatus after the transition

    Raises:
        InvalidTransitionError: If the transition is not valid for the current status
    """
    allowed = _VALID_TRANSITIONS.get(current_status, {})
    if transition not in allowed:
        raise InvalidTransitionError(
            current_status=current_status.value,
            transition=transition.value,
        )
    return allowed[transition]


class StatusUpdateRequest(BaseModel):
    """Input model for updating action status (Story 2.4, AC #5).

    Attributes:
        transition: The status transition to apply (publish, disable, enable)
    """
    transition: StatusTransition


class ActionListItem(BaseModel):
    """Output model for action in admin dashboard list (Story 2.4, AC #2; Story 2.6 tags).

    Lightweight model for listing actions with execution stats and tags.
    Story 2.23: category removed — use tags instead.
    """
    id: int
    name: str
    status: ActionStatus
    engine: ActionEngine
    created_at: datetime
    execution_count: int = 0
    tags: list[str] = Field(default_factory=list)


class PaginationInfo(BaseModel):
    """Pagination metadata for list responses."""
    page: int
    page_size: int
    total_count: int
    total_pages: int


class ActionListResponse(BaseModel):
    """Response model for paginated action list (Story 2.4, AC #2)."""
    data: list[ActionListItem]
    pagination: PaginationInfo | None = None

File: idp-portal/frontend/src/types/api/catalog.ts
import type { PaginationInfo } from './common';
import type { RemediationRule } from './remediation';

// === Catalog Action Types (Story 2.1) ===
// Story 2.30: RefCategory type for administrable categories.
export interface RefCategory {
  id: number;
  code: string;
  label: string;
  display_order: number;
  is_active: number;
}

// Story 13.7: These types are now dynamic (loaded from REF_ENGINES and REF_PLATFORMS tables).
// The union types below are kept for backward compatibility, but values come from API.
// Consider using `string` instead for new code.
export type ActionEngine = 'Oracle' | 'SQL Server' | 'DB2' | string;
export type ActionPlatform = 'AAP' | 'Tower' | 'GitHub Actions' | 'Azure DevOps' | 'Terraform' | string;
export type ActionStatus = 'draft' | 'published' | 'disabled';
/** Impact level for actions (Story 2.5, 2.18). */
export type ImpactLevel = 'low' | 'medium' | 'high' | 'critical';
/** Item type: action or workflow (Story 5.7, AC1). */
export type ItemType = 'action' | 'workflow';

export interface ActionCreate {
  name: string;
  description?: string | null;
  /** Story 5.7: item type (action or workflow). Default: action. */
  item_type?: ItemType;
  /** Story 2.30: Category code (optional, validated against REF_CATEGORIES). */
  category?: string | null;
  /** Engine is required for actions, optional for workflows (Story 5.7). */
  engine?: ActionEngine | null;
  /** Platform is required for actions, optional for workflows (Story 5.7). */
  platform?: ActionPlatform | null;
  /** Story 31.1: Integration FK (optional, derived from integration selection). */
  integration_id?: number | null;
  parameters_schema?: Record<string, unknown> | null;
  /** Story 2.18: impact_rules includes criteria field. */
  impact_rules?: Record<string, { level: ImpactLevel; criteria?: string | null }> | null;
  /** Story 2.18 AC5: Default impact level when no rule matches the environment. */
  default_impact_level?: ImpactLevel | null;
  /** Story 3.4 FR12: Markdown documentation for contextual help. */
  documentation_md?: string | null;
  /** Story 31.8: Notification channels configuration. */
  notification_config?: NotificationConfig | null;
  /** Story 28.4: FK to predefined business rule policy. */
  business_rule_policy_id?: number | null;
  /** Story 63.9: FK vers OutputSchema déclaré par l'admin pour cette action. */
  output_schema_id?: number | null;
}

export interface ActionResponse {
  id: number;
  name: string;
  description: string | null;
  /** Story 5.7: item type (action or workflow). */
  item_type: ItemType;
  /** Story 2.30: Category code (nullable). */
  category?: string | null;
  /** Engine (nullable for workflows). */
  engine: ActionEngine | null;
  /** Platform (nullable for workflows). */
  platform: ActionPlatform | null;
  /** Story 31.1: Integration FK (nullable). */
  integration_id?: number | null;
  parameters_schema: Record<string, unknown> | null;
  /** Story 2.18: impact_rules includes criteria field. */
  impact_rules: Record<string, { level: ImpactLevel; criteria?: string | null }> | null;
  /** Story 2.18 AC5: Default impact level when no rule matches the environment. */
  default_impact_level: ImpactLevel | null;
  /** Story 2.24: change_model_code removed from action level. */
  status: ActionStatus;
  created_by: number | null;
  created_at: string;
  updated_at: string | null;
  tags?: string[];
  /** Story 3.4 FR12: Markdown documentation for contextual help. */
  documentation_md?: string | null;
  /**
   * Story 13.2, AC3: Whether action requires target selection (default true).
   * Story 22.18: MED-6 fix (code-quality-assessment-2026-02-08.md:442-445).
   * Optional because frontend uses defensive logic (!== false) to assume true by default,
   * consistent with backend models.BooleanField(default=True).
   */
  requires_target?: boolean;
  /** Story 64.13: IaC drift tracking. */
  last_synced_at?: string | null;
  last_synced_hash?: string | null;
}

/** Story 31.8: Notification channel configuration. */
export interface NotificationChannel {
  type: 'email' | 'teams' | 'page_oncall';
  enabled: boolean;
  conditions: ('on_failure' | 'on_success' | 'always' | 'on_approval_required')[];
  /** Email only: "requester" or direct email address. */
  recipient?: string;
  /** Teams only: webhook URL or Vault reference. */
  webhook_url_ref?: string;
  /** Page on-call only: API URL (optional, falls back to settings). */
  api_url?: string;
}

/** Story 31.8: Notification configuration for actions. */
export interface NotificationConfig {
  channels: NotificationChannel[];
  page_individual_enabled: boolean;
}

// ADR-007 Story 57.13 — Step types for step-based workflows
// Story 57.16: Schedule step types
/** Source for scheduling date in a schedule_execution step. */
export type ScheduleSource = 'parameter' | 'fixed_offset' | 'recurring';

/** Configuration for a schedule_execution workflow step. */
export interface ScheduleStepConfig {
  /** Source de la date de planification. */
  schedule_source: ScheduleSource;
  /** Nom du paramètre contenant la date ISO 8601 (quand source='parameter'). */
  schedule_parameter_name?: string;
  /** Offset relatif à now() (quand source='fixed_offset'). Format: +Nd, +Nh, +Nw, +Nm. */
  fixed_offset?: string;
  /** Pattern récurrent (quand source='recurring'). */
  recurring_pattern?: {
    pattern_type: 'daily' | 'weekly' | 'cron';
    pattern_config: Record<string, unknown>;
  };
  /** Si true, hérite les paramètres de l'exécution courante (défaut: false). */
  inherit_parameters?: boolean;
  /** Si true, hérite les targets de l'exécution courante (défaut: false). */
  inherit_targets?: boolean;
  /** Mapping de paramètres additionnels. Clé = nom du param cible, Valeur = JSONPath ou valeur statique. */
  parameter_mapping?: Record<string, string>;
}

/** @deprecated parallel_group: supprimé du builder (Story 67-5), conservé pour rétro-compat exécutions. Sera supprimé en Story 67.7. */
export type WorkflowStepType = 'platform' | 'service_call' | 'http_request' | 'evaluation' | 'gate' | 'schedule_execution' | 'parallel_group';

/** Workflow step - reference to an existing action or special step type (Story 5.7, AC2; Story 16.2 branches & retry; Story 57.13 step types). */
export interface WorkflowStep {
  order: number;
  name: string | null;
  /** Story 57.13: Step type. Defaults to 'platform' for backward compatibility. */
  step_type?: WorkflowStepType;
  /** Story 18.3: Real action name from backend (resolved from referenced_action). */
  action_name?: string | null;
  /** Story 16.2: Stable ID for referencing in branches (optional for backward compatibility). */
  step_id?: string | null;
  /** Story 67.4: Multiple next step IDs on success (fan-out). */
  on_success_step_ids?: string[] | null;
  /** Story 67.4: Multiple next step IDs on error (fan-out). */
  on_error_step_ids?: string[] | null;
  /** Story 67.4: Join policy for convergence steps. */
  join_policy?: 'all_success' | 'one_success' | 'all_done' | 'all_failed' | 'one_failed' | null;
  // === platform ===
  /** Story 57.13: Optional for backward compat — required if step_type='platform'. */
  referenced_action_id?: number | null;
  /** Story 16.2: Enable retry for this step (platform only). */
  retry_enabled?: boolean;
  /** Story 16.2: Max retry attempts (default: 3). */
  retry_max_attempts?: number | null;
  /** Story 16.2: Retry interval in seconds (default: 60). */
  retry_interval_seconds?: number | null;
  /** Story 16.2: Backoff multiplier for exponential backoff (default: 2.0). */
  retry_backoff_multiplier?: number | null;
  // === service_call ===
  /** Story 57.13: Integration type for service_call steps ('servicenow' | 'vault' | 'jira'). */
  integration_type?: string | null;
  /** Story 57.13: Operation for service_call steps. */
  operation?: string | null;
  // === evaluation ===
  /** Story 57.13: Business rule policy ID for evaluation steps. */
  policy_id?: number | null;
  // === gate ===
  /** Story 57.13: Gate type for gate steps. */
  gate_type?: 'maintenance_window' | 'approval' | null;
  /** Story 57.13: Action on timeout for gate steps. */
  on_timeout?: 'FAIL' | 'SKIP' | null;
  /** Story 57.13: Step IDs to show to approver (gate type=approval only). */
  context_from?: string[] | null;
  /** Story 58.4: Profile IDs allowed to approve (gate type=approval only). Empty/null = all is_approver profiles eligible. */
  approver_profile_ids?: number[] | null;
  /** Story 57.13: Timeout duration string e.g. '24h', '30m'. */
  timeout?: string | null;
  // === http_request ===
  /** Story 57.13: URL for http_request steps. */
  url?: string | null;
  /** Story 57.13: HTTP method for http_request steps. */
  method?: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | null;
  /** Story 57.13: HTTP headers for http_request steps. */
  headers?: Record<string, string> | null;
  /** Story 57.13: Request timeout in seconds for http_request steps. */
  request_timeout?: number | null;
  // === schedule_execution ===
  /** Story 57.16: Configuration pour les steps de planification. */
  schedule_config?: ScheduleStepConfig | null;
  // === parallel_group — TODO Story 67.7: supprimer ces champs après migration ===
  /** Story 65.4: Liste des step_id à exécuter en parallèle (≥2 requis). */
  parallel_steps?: string[] | null;
  /** Story 65.4: Step suivant si tous les sous-steps réussissent. */
  on_all_success_step_id?: string | null;
  /** Story 65.4: Step suivant si au moins un sous-step échoue (fail-fast). */
  on_any_error_step_id?: string | null;
  // === shared (service_call, evaluation, http_request) ===
  /** Story 57.13: Condition for environment filtering. */
  condition?: { environment_in?: string[] } | null;
  /** Story 57.13: Input mapping (key → value expression). */
  input_mapping?: Record<string, unknown> | null;
  /** Story 57.13: Output mapping (key → JSONPath expression). */
  output_mapping?: Record<string, string> | null;
}

/** Request to update workflow steps (Story 5.7, AC2). */
export interface WorkflowStepsUpdate {
  steps: WorkflowStep[];
}

// === Execution Steps Types (Story 2.2; Story 2.7 connector_type) ===

export type ExecutionStepType = 'prerequisite' | 'execution' | 'verification';

/** Connector type for execution steps (Story 2.7). Aligned with backend ConnectorType. */
export type ConnectorType =
  | 'aap'
  | 'servicenow'
  | 'azuredevops'
  | 'jira'
  | 'github_actions'
  | 'terraform'
  | 'none';

export interface ExecutionStep {
  order: number;
  name: string;
  type: ExecutionStepType;
  connector_type: ConnectorType;
  connector_config?: Record<string, unknown> | null;
  conditional_environments: string[] | null;
}

export interface ExecutionStepsUpdate {
  steps: ExecutionStep[];
}

/** Story 28.1: Business rule policy criteria for review_if_modified. */
export interface BusinessRuleCriteria {
  resource_type?: string;
  attribute_paths?: string[];
}

/** Story 28.1: Business rule policy definition. */
export interface BusinessRulePolicy {
  when: {
    step_type: string;
    output_key?: string;
  };
  policy: {
    type: string;
    require_review_if_modified?: BusinessRuleCriteria[];
    auto_approve_if_none_match?: boolean;
  };
}

/** Story 28.1: Business rule policies JSON structure. */
export interface BusinessRulePoliciesData {
  on_step_output: BusinessRulePolicy[];
}

export interface ActionDetail extends ActionResponse {
  /** Story 2.14: rbac_policies removed — RBAC now managed via profiles. */
  execution_steps: ExecutionStep[] | null;
  /** Story 5.7: workflow steps for workflows (action references). */
  workflow_steps: WorkflowStep[] | null;
  /** Story 31.8: Notification channels configuration. */
  notification_config?: NotificationConfig | null;
  /* documentation_md is inherited from ActionResponse (Story 3.4) */
  /** Story 9.1: Remediation rules for auto-suggesting corrective actions. */
  remediation_rules?: RemediationRule[] | null;
  /** Story 28.1: Business rule policies evaluated on step output (inline legacy). */
  business_rule_policies?: BusinessRulePoliciesData | null;
  /** Story 28.4: FK to predefined business rule policy. */
  business_rule_policy_id?: number | null;
  /** Story 28.4: Name of the predefined business rule policy (computed). */
  business_rule_policy_name?: string | null;
  /** Story 63.9: FK vers OutputSchema déclaré par l'admin pour cette action. */
  output_schema_id?: number | null;
}

// === Status Transition Types (Story 2.4) ===
// Note: Story 2.3 RBAC by action types (UserProfileType, EnvironmentPermission, RbacPolicies,
// RbacPoliciesUpdate) removed in Story 2.14 — RBAC now managed via profiles.

export type StatusTransition = 'publish' | 'disable' | 'enable';

export interface StatusUpdateRequest {
  transition: StatusTransition;
}

export interface ActionListItem {
  id: number;
  name: string;
  /** Story 5.7: item type for workflow icon display. */
  item_type?: ItemType;
  /** Story 2.30: Category code (nullable). */
  category?: string | null;
  status: ActionStatus;
  /** Engine (nullable for workflows). */
  engine: ActionEngine | null;
  /** Story 57.13: Platform for grouping in ActionPalette (returned by eligible-for-workflow via ActionSerializer). */
  platform?: string | null;
  created_at: string;
  execution_count: number;
  tags?: string[];
  /** Story 18.1: soft-delete fields (present when include_disabled=true). */
  deleted_at?: string | null;
  deleted_by?: number | null;
  deletion_reason?: string | null;
  /** Story 22.18: MED-6 fix — Whether action requires target selection (default true). */
  requires_target?: boolean;
  /** Story 64.13: IaC drift tracking. */
  last_synced_at?: string | null;
  last_synced_hash?: string | null;
  updated_at?: string | null;
  /** Story 63.9: FK vers OutputSchema déclaré par l'admin pour cette action (lecture seule). */
  output_schema_id?: number | null;
}

// Story 57.13: BusinessRulePolicyListItem is defined in business_rules.ts (already exists)

export interface AdminActionsFilters {
  status?: ActionStatus;
  engine?: ActionEngine;
  /** Story 5.7: filter by item type (action or workflow). */
  item_type?: ItemType;
  page?: number;
  page_size?: number;
  /** Story 18.1 (AC4/AC5): include disabled actions. Default false. */
  include_disabled?: boolean;
}

export interface ActionListResponse {
  data: ActionListItem[];
  pagination: PaginationInfo | null;
}

// === Preview Types (Story 2.5) ===
// Note: ImpactLevel moved to top of file (before ActionCreate) for Story 2.18.

// === Impact Rules Editor (Story 2.18) ===

/** Single impact rule definition for the visual editor (AC1–AC6). */
export interface ImpactRuleDefinition {
  /** Optional stable id for React key; assigned when adding or loading. */
  id?: string;
  /** Environment name: DEV, STAGING, PROD, etc. */
  environment: string;
  /** Impact level for this environment. */
  level: ImpactLevel;
  /** Justification / criteria for this impact level. */
  criteria: string | null;
}

// === Parameters Editor (Story 2.17) ===

/** JSON Schema type / format for a single parameter. Aligns with backend validate_parameters_schema. */
export type ParameterSchemaType =
  | 'string'
  | 'number'
  | 'integer'
  | 'boolean'
  | 'date'
  | 'date-time'
  | 'select';

/** Inventory source type for parameter fields (Story 23.5). */
export type InventorySourceType = 'servers' | 'instances' | 'databases';

/** Single parameter definition for the visual editor (AC2). */
export interface ParameterDefinition {
  /** Optional stable id for React key / DnD; assigned when adding or loading. */
  id?: string;
  name: string;
  type: ParameterSchemaType;
  required: boolean;
  default?: string;
  description?: string;
  /** When type is 'select', list of enum options. */
  enum?: string[];
  /** Story 23.5: Parameter value source — manual input or inventory selection. */
  source?: 'manual' | 'inventory';
  /** Story 23.5: Inventory entity type when source is 'inventory'. */
  inventory_type?: InventorySourceType;
  /** Story 37.5: Column used as value/label in inventory dropdown (optional). */
  inventory_value_column?: string;
}

/**
 * Subset of ActionDetail used for real-time preview in admin form.
 * Contains all fields needed to render ActionCard and ActionDrawerPreview.
 * Story 2.23: category removed — use tags for categorization.
 */
/** Story 69.2: Included action summary for workflow cards. */
export interface IncludedAction {
  id: number;
  name: string;
  engine: ActionEngine | null;
  parameters_schema?: Record<string, unknown> | null;
}

export interface ActionPreviewData {
  name: string;
  description: string | null;
  /** Story 5.7: item type for workflow icon display. */
  item_type?: ItemType;
  engine: ActionEngine | null;
  platform: ActionPlatform | null;
  impact_level: ImpactLevel | null;
  parameters_schema: Record<string, unknown> | null;
  tags?: string[];
  /** Nombre d'exécutions (affiché si disponible, Task 1.1). */
  execution_count?: number | null;
  /** Story 3.4 FR12: Markdown documentation for contextual help. */
  documentation_md?: string | null;
  /** Story 8.1: Performance stats for action scorecard (optional, loaded separately). */
  stats?: ActionStats | null;
  /** Story 22.18: MED-6 fix — Whether action requires target selection (default true). */
  requires_target?: boolean;
  /** Story 69.2: List of technology engine codes for workflow cards. */
  technologies?: string[];
  /** Story 69.2: Included actions summary for workflow cards (prep for 69.3/69.4). */
  included_actions?: IncludedAction[];
}

/**
 * Action performance statistics (Story 8.1, AC1, AC4).
 * Aggregated metrics over the last 30 days.
 */
export interface ActionStats {
  /** Success rate percentage (COMPLETED / (COMPLETED + FAILED) * 100). Null if no finished executions. */
  success_rate: number | null;
  /** Average execution time in milliseconds for COMPLETED executions. Null if no completed executions. */
  avg_execution_time_ms: number | null;
  /** Total number of executions in the period. */
  total_executions: number;
  /** Number of FAILED executions (incidents). */
  incidents_count: number;
}
